### PR TITLE
Version of the parser from bumpreqs

### PR DESCRIPTION
**Pull request by thatch**
Lots more tests, and keep better track of whitespace

I looked at the version from nflx-advice and it only has an (unreferenced) `UrlRequirement` class.  After this gets merged can you try to figure out what we'd intended to do with that?

### CODE REVIEW

File: codemod_pip_requirements/__main__.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import sys
+from . import RequirementFile
+
+if __name__ == "__main__":
+
```
Comment (jag250) [inline codemod_pip_requirements/__main__.py]: Is this entry point for testing? Also there's an unnecessary newline here

File: codemod_pip_requirements/core.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        buf.write(self.whitespace_after_requirement)
+        buf.write(self.comment)
+        buf.write(self.newline)
+
+    def dump(self, buf: TextIO, indent: str = "") -> None:
+        buf.write(indent + repr(self) + "\n")
+
+
+WELL_FORMED_REQUIREMENT_LINE = re.compile(
+    r"([ \t]*)([^# \t][^#]*?)([ \t]*)(#.*?)?(\r?\n)$"
```
Comment (jag250) [inline codemod_pip_requirements/core.py:77]: Is this taken from a PEP? (Or PyPA specification)

File: codemod_pip_requirements/core.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    def build(self, buf: TextIO) -> None:
+        buf.write(self.orig_line)
+
+    def dump(self, buf: TextIO, indent: str = "") -> None:
+        buf.write(indent + repr(self) + "\n")
+
+
+@dataclass
+class VcsRequirementLine(BaseNode):
+    requirement: str
```
Comment (jag250) [inline codemod_pip_requirements/core.py:39]: Does `requirement` in this class mean something like `ick @ https://github.com/advice-animal/ick.git` including all the whitespace between `ick`, `@`, and the github link?

File: codemod_pip_requirements/core.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    def build(self, buf: TextIO) -> None:
+        buf.write(self.requirement)
+        buf.write(self.whitespace_before_requirement)
+        buf.write(self.whitespace_after_requirement)
+        buf.write(self.comment)
+        buf.write(self.newline)
+
+    def dump(self, buf: TextIO, indent: str = "") -> None:
+        buf.write(indent + repr(self) + "\n")
```
Comment (jag250) [inline codemod_pip_requirements/core.py]: I find it interesting that `VcsRequirementLine`, `LocalPathRequirementLine`, and `RequirementLine` all have the same member data and methods. Are there plans to add more methods specific to each type of line? If not, I would maybe make these one class and add a `type` data member specifying what kind of line it is.

File: codemod_pip_requirements/core.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+@dataclass
+class RequirementFile:
+    children: Sequence[BaseNode] = ()
+
+    @classmethod
+    def parse(cls, data: str) -> "RequirementFile":
+        children: list[BaseNode] = []
+        if not data.endswith("\n"):
+            if "\r\n" in data:
```
Comment (jag250) [inline codemod_pip_requirements/core.py:91]: I remember from `advice-animal` that this should be for Windows. I think there should be a comment explaining that.

File: codemod_pip_requirements/core.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+            (
+                whitespace_before_requirement,
+                value,
+                whitespace_after_requirement,
+                comment,
+                newline,
+            ) = match.groups()
+
+            if value.startswith(("-", "#")):
```
Comment (jag250) [inline codemod_pip_requirements/core.py:110]: Would this parse `-e ../foo` (the comment on line 57) the way you want it to?

File: codemod_pip_requirements/core.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                        whitespace_after_requirement=whitespace_after_requirement,
+                        comment=comment or "",
+                        newline=newline,
+                    )
+                )
+                continue
+            # N.b. see COMMENT_RE in pip/req/req_file.py
+            elif value and comment and not whitespace_after_requirement:
+                # TODO strip
+                value = value + whitespace_after_requirement + comment
```
Comment (jag250) [inline codemod_pip_requirements/core.py:127]: Assuming `whitespace_after_requirement` is falsey, why is it here?

(Not too sure what the #TODO is referring to, sorry if you already had this in mind)

File: codemod_pip_requirements/core.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+                        requirement=value,
+                        whitespace_before_requirement=whitespace_before_requirement,
+                        whitespace_after_requirement=whitespace_after_requirement,
+                        comment=comment or "",
+                        newline=newline,
+                    )
+                )
+                continue
+
+            Requirement(value)  # validates
```
Comment (jag250) [inline codemod_pip_requirements/core.py:148]: Do you want error handling here? I guess `packaging.requirements.InvalidRequirement` is pretty readable as to what the problem is.

File: codemod_pip_requirements/core.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+                value = value + whitespace_after_requirement + comment
+                children.append(
+                    VcsRequirementLine(
+                        requirement=value,
+                        whitespace_before_requirement=whitespace_before_requirement,
+                        newline=newline,
+                    )
+                )
+                continue
+            elif "://" in value:
```
Comment (jag250) [inline codemod_pip_requirements/core.py:136]: Depending on how thorough you want to be, iirc `file:///absolute/path/to/package-1.0.0-py3-none-any.whl` is valid. (thanks chatgpt for the example)

File: codemod_pip_requirements/core.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+@dataclass
+class BaseNode:
+    def with_changes(self: T, **kwargs: Any) -> T:
+        return replace(self, **kwargs)  # type: ignore
+
+    def build(self, buf: TextIO) -> None:
+        raise NotImplementedError
+
+    def dump(self, buf: TextIO, indent: str = "") -> None:
```
Comment (jag250) [inline codemod_pip_requirements/core.py:20]: I'm curious as to the difference in intended use between the `.build()` and `.dump()` methods. A docstring could be useful here.

File: codemod_pip_requirements/core.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+from dataclasses import dataclass, replace
+from typing import Sequence, Any, TypeVar, TextIO
+
+from packaging.requirements import Requirement
+
+T = TypeVar("T")
+
+
+@dataclass
+class BaseNode:
```
Comment (jag250) [inline codemod_pip_requirements/core.py:13]: This name wasn't super clear to me. Maybe add a docstring saying that it exists for polymorphism in `RequirementFile.children`? (I also have zero experience with parsers so take this with a shaker of salt)

File: codemod_pip_requirements/__init__.py [hunk 12]
```diff
@@ -2,3 +2,5 @@
     from ._version import __version__
 except ImportError:  # pragma: no cover
     __version__ = "dev"
+
+from .core import *  # noqa: F403
```
Comment (jag250) [inline codemod_pip_requirements/__init__.py:6]: Why is using `*` necessary?

File: codemod_pip_requirements/core.py (hunk 3 above)
Comment (jag250) [reply codemod_pip_requirements/core.py:39]: Reading the tests answered this question, nvm.

File: tests/test_cst.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+
+def test_special_case_url_hash_mark():
+    expected = "http://foo#egg_info=t\n"
+    r = RequirementFile.parse(expected)
+    assert len(r.children) == 1
+    assert isinstance(r.children[0], VcsRequirementLine)
+    assert r.children[0].requirement == "http://foo#egg_info=t"
+    t = StringIO()
+    r.build(t)
```
Comment (jag250) [inline tests/test_cst.py:139]: nit: I didn't see any usage of `.dump()` in the tests.

File: codemod_pip_requirements/__main__.py (hunk 1 above)
Comment (thatch) [reply codemod_pip_requirements/__main__.py]: I'd call it a "demo" more than testing, although it's uncovered.  Will fix newline.

File: codemod_pip_requirements/__init__.py (hunk 12 above)
Comment (thatch) [reply codemod_pip_requirements/__init__.py:6]: Laziness; better would be specifying the names explicitly here and also putting them in `__all__`.  Let's resolve the UrlRequirementLine debate (that's closer to "structural" because it also affects what gets written here)

File: codemod_pip_requirements/core.py (hunk 11 above)
Comment (thatch) [reply codemod_pip_requirements/core.py:13]: Naming suggestion?  I might also make a ParsedLine intermediate class to reduce duplication with build etc

File: codemod_pip_requirements/core.py (hunk 10 above)
Comment (thatch) [reply codemod_pip_requirements/core.py:20]: Dump is for debugging.  Agreed.

File: codemod_pip_requirements/core.py (hunk 3 above)
Comment (thatch) [reply codemod_pip_requirements/core.py:39]: Yes.  The idea is that the caller could call `packaging.requirements.Requirement(requirement)` on this string.

See https://github.com/pypa/packaging/blob/14b83e15dbb9caa87c63646ba7808b2b5e460ce6/src/packaging/_parser.py#L50-L55

(That should be in a comment/docstring, will do)

File: codemod_pip_requirements/core.py (hunk 4 above)
Comment (thatch) [reply codemod_pip_requirements/core.py]: I was thinking the same thing, that's the "ParsedLine intermediate class" mentioned in a pr comment above

File: codemod_pip_requirements/core.py (hunk 6 above)
Comment (thatch) [reply codemod_pip_requirements/core.py:110]: Nope.  Good catch

File: codemod_pip_requirements/core.py (hunk 9 above)
Comment (thatch) [reply codemod_pip_requirements/core.py:136]: I haven't decided how different Vcs and Url support should be.  Needs more test data.

**Comment by jag250** [comment]
> Lots more tests, and keep better track of whitespace
> 
> I looked at the version from nflx-advice and it only has an (unreferenced) `UrlRequirement` class. After this gets merged can you try to figure out what we'd intended to do with that?

IIRC the `UrlRequirement` was for something that was online in a remote archive but isn't a VCS. Something like `"my-package @ https://example.com/my-package-0.1.0.tar.gz"` ([this example is from the poetry docs](https://python-poetry.org/docs/dependency-specification/#url-dependencies)). 

A real-world example could be the index on pytorch.org: `https://download.pytorch.org/whl/cpu-cxx11-abi/torch-2.0.1%2Bcpu.cxx11.abi-cp311-cp311-linux_x86_64.whl`.

**Comment by thatch** [comment]
Mostly addressed comments; Jadon says ok to merge and get a release out.
